File: python/find_the_index_of_the_first_occurrence_in_a_string.py
class Solution:
    # AAABAAAA
    # AAAA
    # 0123
    # NOTE: O(m+n)
    def strStr(self, haystack: str, needle: str) -> int:
        if len(needle) == 0:
            return 0

        lps: list[int] = [0] * len(needle)
        i: int = 1
        preLPS: int = 0
        while i < len(needle):
            if needle[preLPS] == needle[i]:
                preLPS += 1
                lps[i] = preLPS
                i += 1
            elif preLPS == 0:
                lps[i] = 0
                i += 1
            else:
                preLPS = lps[preLPS - 1]

        m: int = 0
        n: int = 0
        while m < len(haystack):
            if haystack[m] == needle[n]:
                m += 1
                n += 1
            else:
                if n == 0:
                    m += 1
                else:
                    n = lps[n - 1]

            if n == len(needle):
                return m - n

        return -1


def test_strStr_case_1():
    # arrange
    haystack: str = "sadbutsad"
    needle: str = "sad"
    expected: int = 0

    # act
    solution = Solution()
    actual = solution.strStr(haystack, needle)

    # assert
    assert expected == actual


def test_strStr_case_2():
    # arrange
    haystack: str = "leetcode"
    needle: str = "leeto"
    expected: int = -1

    # act
    solution = Solution()
    actual = solution.strStr(haystack, needle)

    # assert
    assert expected == actual


def test_strStr_case_3():
    # arrange
    haystack: str = "mississippi"
    needle: str = "issip"
    expected: int = 4

    # act
    solution = Solution()
    actual = solution.strStr(haystack, needle)

    # assert
    assert expected == actual

File: rust/src/code/find_the_index_of_the_first_occurrence_in_a_string.rs
struct Solution {}

impl Solution {
    // AAABAAAA
    // AAAA
    // 0123
    pub fn str_str(haystack: String, needle: String) -> i32 {
        if needle.len() == 0 {
            return 0;
        }

        let needle: Vec<char> = needle.chars().collect();
        let haystack: Vec<char> = haystack.chars().collect();
        let mut lps: Vec<usize> = vec![0; needle.len()];
        let mut i: usize = 1;
        let mut pre_lps: usize = 0;

        while i < needle.len() {
            if needle[pre_lps] == needle[i] {
                pre_lps += 1;
                lps[i] = pre_lps;
                i += 1;
            } else if pre_lps == 0 {
                lps[i] = 0;
                i += 1;
            } else {
                pre_lps = lps[pre_lps - 1];
            }
        }

        let mut m: usize = 0;
        let mut n: usize = 0;
        while m < haystack.len() {
            if haystack[m] == needle[n] {
                m += 1;
                n += 1;
            } else {
                if n == 0 {
                    m += 1;
                } else {
                    n = lps[n - 1];
                }
            }

            if n == needle.len() {
                return (m - n) as i32;
            }
        }

        -1
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_str_str_case_1() {
        // arrange
        let haystack: String = "sadbutsad".to_string();
        let needle: String = "sad".to_string();
        let expected: i32 = 0;

        // act
        let mut actual = Solution::str_str(haystack, needle);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_str_str_case_2() {
        // arrange
        let haystack: String = "leetcode".to_string();
        let needle: String = "leeto".to_string();
        let expected: i32 = -1;

        // act
        let mut actual = Solution::str_str(haystack, needle);

        // assert
        assert_eq!(expected, actual);
    }
}
